docs/shopify_app/sessions.md

```diff
@@ -203,6 +204,66 @@ user.with_shopify_session do
 end
 ```
 
+#### Re-fetching an access token when API returns Unauthorized
+
+When using `ShopifyApp::EnsureHasSession` and the `new_embedded_auth_strategy` configuration, any **unhandled** Unauthorized `ShopifyAPI::Errors::HttpResponseError` will cause the app to perform token exchange to fetch a new access token from Shopify and the action to be retried once.
+
+```ruby
+class MyController < ApplicationController
+  include ShopifyApp::EnsureHasSession
+
+  def index
+    client = ShopifyAPI::Clients::Graphql::Admin.new(session: current_shopify_session)
+    client.query(options) # when this raises an Unauthorized error from Shopify, EnsureHasSession will retry the action once after performing token exchange
+  end
+end
+```
+
+If the error is being rescued in the action, it's still possible to perform token exchange and retry again with the `with_token_refetch` method provided by `EnsureHasSession`.
+
+```ruby
+class MyController < ApplicationController
+  include ShopifyApp::EnsureHasSession
+
+  def index
+    client = ShopifyAPI::Clients::Graphql::Admin.new(session: current_shopify_session)
+    with_token_refetch(current_shopify_session, session_token) do # Unauthorized errors raised in this block will initiate token exchange and the block will be retried once
```

Comment:
hmm looking at the example here, do you think it might be confusing to devs on the difference between `current_shopify_session` and `session_token`? There are so many cases where we just use "session" to describe something, should we rename `session_token` to be like `session_jwt_token` to be more explicit on which one we're talking about?

Reply:
Yeah I agree it's very confusing. I noticed we call it `id_token` in the URL and also in [App Bridge](https://shopify.dev/docs/api/app-bridge-library/apis/id-token), so I'm leaning towards `id_token` here to keep consistence, especially because those are the two sources of the token we're using here.